pkg/datastore/firestore/iterator.go

```diff
@@ -33,10 +36,23 @@ func (it *Iterator) Next(dst interface{}) error {
 		}
 		return err
 	}
+
+	// Update last iterated item as last read doc.
+	it.last = doc
+
 	return doc.DataTo(dst)
 }
 
 func (it *Iterator) Cursor() (string, error) {
-	// Note: Cursor function is not needed in Cloud Firestore.
-	return "", datastore.ErrUnimplemented
+	if it.last == nil {
+		return "", datastore.ErrInvalidCursor
+	}
+
+	cursorVal, err := json.Marshal(it.last)
+	if err != nil {
+		return "", err
+	}
+
+	// TODO: Encrypt cursor string value before pass it to the caller.
```

Comment:
Because a TODO comment, it obviously should be mentioned at that time, but I'm curious about why you're looking to encrypt it. I would say it's better to utilize the lowest cost and recoverable method possible, such as base64. Sorry if this is just a temporary word choice.

Reply:
No wonder, any comment is welcomed. 😁 We want to avoid leak unnecessary fields to the frontend as much as possible but keep the building logic as easy as possible. At this point, we don't know the kind of object here such that our choices for making a string out of an object is only toString or JSON marshal, and in both cases, every fields of that object will be shown to the caller side (and keep it as is up to the frontend). To avoid the security issue (I guess), we should make it not easy to retrieve by encrypting this value before passing it 👀

Reply:
That makes sense 😃  I totally slipped my mind that the Cursor will be passed to the web frontend. Nailed it.

Reply:
(test for slack)

Reply:
@nakabonne as the idea shared by nghialv in https://github.com/pipe-cd/pipe/pull/1864#discussion_r613751694 👀 
I reimplemented this part, the cursor contains only fields listed in the ordering list, so that we don't have to encrypt the cursor since not all the fields but just specified (ordering fields) are exposed to the client.
addressed on https://github.com/pipe-cd/pipe/pull/1864/commits/9653d7e22de54cbf95eae18043854e681fad128e

Reply:
Cool. That's definitely better over encryption entirely. Another thing I'm wondering is why not using just Id? I mean Id is unique, and order is already specified using the `q.OrderBy()` function. So I'm guessing what we need to pass to the `StartAfter()` function is only `Id`, but am I wrong? Any kinds of performance issues exist right?

Reply:
The pagination required result set is sorted in another way than just Id (for example want to list all deployment in order by UpdatedAt) in that case, just Id is not enough to consistent the order of result set when we move the second page.

Reply:
You can read more about this from here: https://cloud.google.com/firestore/docs/query-data/query-cursors 👀

Reply:
I assumed that `q.OrderBy()` is enough to get ordering tasks done, but looks wrong. Okay, thank you for letting me know!

Reply:
> Another thing I'm wondering is why not using just Id? I mean Id is unique, and order is already specified using the q.OrderBy() function.

@nakabonne Just randomly read this comment but I get your point now 😄 Your thought is right in case the result set of sorted entry list items is existed and stored somewhere in the server and on each query, we just pull the part we want from a specified id - in that case just id is enough 👍 But the truth is, datastore does not hold that sorted result set for us, that why we only have 2 options: `using offset to skip to specified id in case we using the data stored order` or `using the value of the field we use to order the data set and query for the part which has values greater/lower than that value. As for Firestore, maybe it supported the first option by stored data as both indexes order and raw data stored order, but to keep up a common interface for other datastores (our MySQL in that case) the other fields than just Id is needed. 😁